### PR TITLE
[refactor-prompt] add token sendfrom

### PR DESCRIPTION
**Pull request by ixje**
**What current issue(s) does this address, or what feature is it adding?**
Moved #623 forward by implementing `sendfrom`

**How did you solve this problem?**
write code + tests

**How did you make sure your solution works?**
add tests + make tests

**Are there any special changes in the code that we should be aware of?**
Yes.
1.  I noticed that many `wallet token <something` have to validate the same parameters (token, from_addr, to_addr, amount) over and over again. I added a helper function and also changed `wallet token send` to use that function.
2. `token_send_from()` used to mix business logic + representation (representation meaning; printing and asking for prompt). I changed it to only do the test sending part of the business logic and not the representation. This allows others to re-use the function more easily.
3. I removed a bunch of tests that are now covered by the specific `wallet token sendfrom`  tests which I wrote before noticing the old tests were broken from changing the `token_send_from()`

**Please check the following, if applicable:**

- [X] Did you add any tests?
- [X] Did you run `make lint`?
- [X] Did you run `make test`?
- [X] Are you making a PR to a feature branch or development rather than master?
- [] Did you add an entry to `CHANGELOG.rst`? (if not, please do)

### CODE REVIEW

File: neo/Prompt/Commands/Tokens.py [hunk 1]
```diff
@@ -131,32 +138,114 @@ def command_desc(self):
...
+        from_addr = arguments[1]
+        to_addr = arguments[2]
+
+        try:
+            amount = float(arguments[3])
+        except ValueError:
+            print(f"{arguments[3]} is not a valid amount")
+            return False
+
+        prompt_passwd = True
```
Comment (jseagrave21) [inline neo/Prompt/Commands/Tokens.py]: When we rebuild `Send.py`, we eliminated this argument.
https://github.com/CityOfZion/neo-python/blob/e3780126f9cf8175f9f3c7ad27f36f2baae20508/neo/Prompt/Commands/Send.py#L165-L169
Is there ever a case where the user should not be prompted? Also, in the case where other programs are using this function, they could use `mock` like I have implemented here: https://github.com/CityOfZion/neo-privatenet-docker/pull/63/files#diff-fb19bbb688a846c3107e299fceb63240R138

I recommend simplifying the code and removing `prompt_passwd` and allow dependent descriptions, etc.

Comment (ixje) [reply neo/Prompt/Commands/Tokens.py]: > Is there ever a case where the user should not be prompted?

Well, if the user is very confident in his command typing he could choose to run something potentially dangerous without confirmation. 

> Also, in the case where other programs are using this function, they could use mock like I have implemented here: https://github.com/CityOfZion/neo-privatenet-docker/pull/63/files#diff-fb19bbb688a846c3107e299fceb63240R138

`mock` is a unit testing library, using it for other purposes is mostly a hack to work around inflexible code. What you've linked kind of proves that the  `process_transaction()` function is inflexible. That function (like many I've recently encountered) mixes business logic with presentation logic (`prompt` actions). The easiest fix for `process_transaction()` would be to add an optional parameter whether to prompt for a password. Then your mocking isn't needed anymore.

Comment (jseagrave21) [reply neo/Prompt/Commands/Tokens.py]: Okay, if that is preferred then I will add `prompt_passwd` back into `Send.py`.

File: neo/Prompt/Commands/Tokens.py [hunk 2]
```diff
@@ -101,16 +108,16 @@ def execute(self, arguments):
...
-        send_from = arguments[1]
-        send_to = arguments[2]
+        from_addr = arguments[1]
+        to_addr = arguments[2]
         try:
             amount = float(arguments[3])
         except ValueError:
             print(f"{arguments[3]} is not a valid amount")
-            return False
+            return
```
[the remark this answers is not in the document]
Comment (ixje) [reply neo/Prompt/Commands/Tokens.py]: nah probably due to merging

File: neo/Prompt/Commands/Tokens.py [hunk 3]
```diff
@@ -314,32 +414,44 @@ def token_approve_allowance(wallet, args, prompt_passwd=True):
...
-    tx, fee, results = token.Allowance(wallet, allowance_from, allowance_to)
+    Raises:
+        ValueError: for invalid arguments or if allowance could not be queried
+
+    Returns:
+        int: allowance
+    """
+    try:
+        placeholder_amount = 0
+        token = _validate_nep5_args(wallet, token_str, from_addr, to_addr, placeholder_amount)
```
[the remark this answers is not in the document]
Comment (ixje) [reply neo/Prompt/Commands/Tokens.py]: yeah much better

**Comment by ixje** [comment]
Thanks both for the review 👍

File: neo/Prompt/Commands/Tokens.py [hunk 4]
```diff
@@ -132,8 +139,90 @@ def command_desc(self):
...
+            # invalid arguments or bad allowance
+            print(str(e))
+            return False
+        except Exception as e:
+            # we act as the final capturing place
+            print("Something really unexpected happened")
+            logger.error(traceback.format_exc())
+            return False
+
+        if tx is not None and results is not None and len(results) > 0:
```
[the remark this answers is not in the document]
Comment (LysanderGG) [reply neo/Prompt/Commands/Tokens.py]: Note that `len(results) > 0` can also be removed because an empty list evaluates to False

**Pull request merged**